File: 第13周/第二题 用分量形式/matlab/jacobi_inv.m
function x = jacobi_inv( A,b )
%JACOBI_INV 用jacobi迭代法解Ax=b

N=size(A,1);


x0=b;
x=zeros(N,1);

for k=1:N
    x(k)=1/A(k,k)*(b(k)-A(k,:)*x0+A(k,k)*x0(k));
end




while(norm(x-x0)>1e-10)
    x0=x;
    for k=1:N
        x(k)=1/A(k,k)*(b(k)-A(k,:)*x0+A(k,k)*x0(k));
    end
end


end

File: 第13周/第二题 用分量形式/python/jacobi_inv.py
import numpy as np
def J_ve(A,b):
    n=np.shape(A)[0]
    xk=np.matrix(np.ones((n,1)))
    xk1=np.matrix(np.zeros((n,1)))
    error=1e-10
    xk=b.copy()
    for i in range(0,n):
        xk1[i,0]=(b[i,0]-A[i,:]*xk[:,0])/A[i,i]+xk[i,0]
    # print(np.linalg.norm(xk-xk1))
    #这怎么标准不统一
    while(np.linalg.norm(xk-xk1)>error):
        xk=xk1.copy();
        for i in range(0,n):
            xk1[i,0]=(b[i,0]-A[i,:]*xk[:,0])/A[i,i]+xk[i,0]
    return xk1